### PR TITLE
Feature/add get departments tree endpoint

### PR DESCRIPTION
**Pull request by AnnPostnikova**

### CODE REVIEW

File: src/DepartmentService.Business/Category/FindCategoryCommand.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        return _responseCreator.CreateFailureFindResponse<CategoryInfo>(HttpStatusCode.BadRequest, errors);
+      }
+
+      FindResultResponse<CategoryInfo> response = new();
+
+      (List<DbCategory> dbCategories, int totalCount) = await _categoryRepository.FindCategoriesAsync(filter);
+
+      response.TotalCount = totalCount;
+
+      response.Body = dbCategories.Select(c => _mapper.Map(c)).ToList();
```
Comment (VLiamin) [inline src/DepartmentService.Business/Category/FindCategoryCommand.cs:50]: may me make easier Select(_mapper.Map)

File: src/DepartmentService.Business/Department/EditDepartmentCommand.cs [hunk 2]
```diff
@@ -26,6 +29,18 @@ public class EditDepartmentCommand : IEditDepartmentCommand
...
     private readonly IResponseCreator _responseCreator;
     private readonly IGlobalCacheRepository _globalCache;
+    private readonly IMemoryCache _cache;
+
+    private void GetArchivedIds(in List<Tuple<Guid, string, string, Guid?>> listDepartments, Guid? idParent, List<Guid> archivedIds)
+    {
+      List<Guid> childrenIds = listDepartments.Where(ld => ld.Item4 == idParent).Select(ld => ld.Item1).ToList();
+      archivedIds.AddRange(childrenIds);
+
+      foreach (var childId in childrenIds)
```
Comment (VLiamin) [inline src/DepartmentService.Business/Department/EditDepartmentCommand.cs:39]: remove var

File: src/DepartmentService.Data/CategoryRepository.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    public CategoryRepository(IDataProvider provider)
+    {
+      _provider = provider;
+    }
+
+    public async Task<Guid?> CreateAsync(DbCategory category)
+    {
+      _provider.Categories.Add(category);
+      await _provider.SaveAsync();
```
Comment (VLiamin) [inline src/DepartmentService.Data/CategoryRepository.cs:26]: remove await

File: src/DepartmentService.Data/DepartmentRepository.cs [hunk 4]
```diff
@@ -179,5 +184,38 @@ public async Task<bool> ExistAsync(Guid departmentId)
...
+      return await departments.Include(x => x.Category).Select(x => new Tuple<Guid, string, string, Guid?>(x.Id, x.Name, x.Category.Name, x.ParentId)).ToListAsync();
+    }
+
+    public async Task RemoveAsync(List<Guid> departmentsIds)
+    {
+      List<DbDepartment> dbDepartments = await _provider.Departments.Where(d => departmentsIds.Contains(d.Id)).ToListAsync();
+
+      dbDepartments.ForEach(x => x.IsActive = false);
+
+      await _provider.SaveAsync();
```
Comment (VLiamin) [inline src/DepartmentService.Data/DepartmentRepository.cs:218]: and here

File: src/DepartmentService.Data/CategoryRepository.cs (hunk 3 above)
Comment (AnnPostnikova) [reply src/DepartmentService.Data/CategoryRepository.cs:26]: why?

File: src/DepartmentService.Data.Provider.MsSql.Ef/Migrations/20211012014055_InitialCreate.cs [hunk 5]
```diff
@@ -37,6 +39,23 @@ private void CreateTableDeparments(MigrationBuilder migrationBuilder)
...
+    private void CreateTableCategories(MigrationBuilder migrationBuilder)
+    {
+      migrationBuilder.CreateTable(
+        name: DbCategory.TableName,
+        columns: table => new
+        {
+          Id = table.Column<Guid>(nullable: false),
+          Name = table.Column<string>(nullable: false),
+          CreatedBy = table.Column<Guid>(nullable: false),
+          CreatedAtUtc = table.Column<DateTime>(nullable: false)
```
Comment (MariyaSin) [inline src/DepartmentService.Data.Provider.MsSql.Ef/Migrations/20211012014055_InitialCreate.cs:51]: add 
          ModifiedBy = table.Column<Guid>(nullable: true),
          ModifiedAtUtc = table.Column<DateTime>(nullable: true)

File: src/DepartmentService.Data/CategoryRepository.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      {
+        query = query.Where(g => g.Name.ToLower().Contains(filter.NameIncludeSubstring.ToLower()));
+      }
+
+      return (
+        await query.Skip(filter.SkipCount).Take(filter.TakeCount).ToListAsync(),
+        await query.CountAsync());
+    }
+
+    public Task<bool> DoesCategoryAlreadyExistAsync(string categoryName)
```
Comment (MariyaSin) [inline src/DepartmentService.Data/CategoryRepository.cs:60]: you in category repo? no need write the word "Category" => DoesNameExistAsync

File: src/DepartmentService.Models.Db/DbCategory.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+namespace LT.DigitalOffice.DepartmentService.Models.Db
+{
+  public class DbCategory
+  {
+    public const string TableName = "Categories";
+
+    public Guid Id { get; set; }
+    public string Name { get; set; }
+    public Guid CreatedBy { get; set; }
+    public DateTime CreatedAtUtc { get; set; }
```
Comment (MariyaSin) [inline src/DepartmentService.Models.Db/DbCategory.cs:15]: add modifed

File: src/DepartmentService.Models.Dto/Models/DepartmentsTreeInfo.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace LT.DigitalOffice.DepartmentService.Models.Dto.Models
+{
+  public class DepartmentsTreeInfo
```
Comment (MariyaSin) [inline src/DepartmentService.Models.Dto/Models/DepartmentsTreeInfo.cs:6]: record

File: src/DepartmentService/Controllers/DepartmentsTreeController.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+{
+  [Route("[controller]")]
+  [ApiController]
+  public class DepartmentsTreeController : ControllerBase
+  {
+    [HttpGet("get")]
+    public async Task<OperationResultResponse<List<DepartmentsTreeInfo>>> GetDepartmentsTreeAsync(
+      [FromServices] IGetDepartmentsTreeCommand command,
+      [FromQuery] FindDepartmentFilter filter
+      )
```
Comment (MariyaSin) [inline src/DepartmentService/Controllers/DepartmentsTreeController.cs:19]: move to 18 line

**Pull request closed without merging**